### ISSUE TITLE
Not receiving foreground or background notifications on ios with Testflight

### ISSUE THREAD

**Issue by michorvath** [invertase/react-native-firebase#4310]
I am rebuilding a pre-existing react native app for a client that used firebase for push notifications. I followed all the docs and believe that I have set up Firebase correctly, however, no message events ever get called. My app successfully request permission for both normal and critical alerts, I get an FCM token, then I copied the token into the Firebase console and attempted to send a test notification and my onMessage never fires off an alert. I don't appear to be receiving any notifications ever.

My package.json contains:
```
{
    "@react-native-firebase/app": "^8.3.0",
    "@react-native-firebase/messaging": "^7.6.1"
}
```

My app is an ejected expo app, not sure if that could be causing issues. The Firebase project was already set up for the clients previous app, so they already had an APNs auth key. The app has push notifications enabled, background modes (fetch, remote notifications, and processing) are all enabled. I added #import <Firebase.h> and [FIRApp Configure]; to my AppDelegate.m as well. Xcode can see my GoogleService-Info.plist and my Info.plist has FirebaseAppDelegateProxyEnabled set to true as one article I found recommended.

My index.js:
```
import { registerRootComponent } from 'expo';
import messaging from '@react-native-firebase/messaging';

import App from './App';

// Register background handler
messaging().setBackgroundMessageHandler(async (remoteMessage) => {
    console.log('Message handled in the background!', remoteMessage);
});

// registerRootComponent calls AppRegistry.registerComponent('main', () => App);
// It also ensures that whether you load the app in the Expo client or in a native build,
// the environment is set up appropriately
registerRootComponent(App);
```

My getFCMToken function (called during signin):
```
export async function getFirebaseToken() {
    try {
        let authStatus = await messaging().hasPermission();

        if (authStatus !== messaging.AuthorizationStatus.AUTHORIZED) {
            authStatus = await messaging().requestPermission({
                sound: true,
                badge: true,
                alert: true,
                carPlay: true,
                announcement: true,
                criticalAlert: true
            });
        }

        let token = await AsyncStorage.getItem('deviceToken');

        if (authStatus === messaging.AuthorizationStatus.AUTHORIZED) {
            if (token) {
                return token;
            }

            token = await messaging().getToken();
            await AsyncStorage.setItem('deviceToken', token);
            return token;
        } else {
            if (token) {
                await AsyncStorage.removeItem('deviceToken');
                messaging().deleteToken(token);
            }

            return false;
        }
    } catch (e) {
        Alert.alert('failed to get FCM token ' + JSON.stringify(e));
        return false;
    }
}
```

My onMessage looks like this:
```
        this.onMessageListener = messaging().onMessage(async (remoteMessage) => {
            Alert.alert('remote message received ' + JSON.stringify(remoteMessage));

            let {notification} = remoteMessage;

            await notifee.displayNotification({
                id: notification.notificationId,
                title: notification.title,
                subtitle: notification.subtitle,
                body: notification.body,
                data: notification.data,
                android: {
                    channelId: channelId,
                    sound: 'default',
                    importance: AndroidImportance.HIGH
                },
                ios: {
                    critical: true,
                    sound: 'default',
                    foregroundPresentationOptions: {
                        alert: true,
                        badge: true,
                        sound: true
                    }
                }
            });
    });
```

I've seen a lot of articles and such with similar issues that usually ended with no solution, or, with a solution for an outdated version of Firebase. I haven't been able to figure out what might be the cause of the issue. I am testing from a real device and not the ios simulator but I am publishing to Testflight, which I'm not sure if that could be causing a problem if the profiles/certs might not be right? The app archives and signs correctly so I don't know if that could actually be a thing.

Any help would be greatly appreciated! I can post additional code and information if necessary.

**Comment by mikehardy** [comment]
For successful troubleshooting you will want to use the REST FCM endpoint to send hand-crafted JSON so you know with absolute certainty what you are sending

For data-only messages you must send content-available: 1

With a notification payload it will pop the notification and I think different APIs get called - but you'll want a full-featured notification package anyway and it will handle the taps from those.

I don't know about old articles and what-not but the issues tracker here has some definitive comments 

I would recommend maybe trying a clean bare-bones project that was focused solely on getting these APIs to work so you had the equivalent of a test lab. These APIs are hard enough to get to work that if I were you I would want to isolate old code first if I were having problems - you can use this to make a skeleton: https://github.com/mikehardy/rnfbdemo/blob/master/make-demo.sh

**Comment by michorvath** [comment]
Thanks @mikehardy for responding!

I generated an OAuth2 token using node and the private key setup explained in the docs so I could send a RESTful test notification via Postman, as you described. My raw body currently looks like this:

```
{   
    "validate_only": true,
    "message": {
        "token": "cImJx3LYv04-q_nlKQewpV:APA91b-GLK9XlmIe3ZVcoHg2Rt-X56E9MBZzuwWnWNnoWi-hnlnvA2kAB3pHu7sY0f-UGpXYzaRDKyGNgxW-mh9Bj4cbiNIRaqeuYQep-GKRy3WP5KWTuwDP85mN-hjAU1uj3srWSNLeUIPdjP",
        "data": {
            "title": "Test Notification",
            "body": "Hello world.",
            "alert_type": "1",
            "android_channel_id": "catapult_channel"
        },
        "android": {
            "priority": 10
        },
        "apns": {
            "payload": {
                "aps": {
                    "alert": {
                        "alert": "Test Notification",
                        "body": "Hello world.",
                        "content_available": 1
                    },
                    "sound": {
                        "critical": 1,
                        "name": "",
                        "volume": 1
                    }
                }
            }
        }
    }
}
```

And I receive a 400 back with the message "The registration token is not a valid FCM registration token". The FCM token I am using is from an actually iOS device and uses the code in my first post. The app id, bundle id, etc is all correct in the project so I am lost on why I am getting a token back but it is invalid. I am looking through the docs and other posts as to not take up too much of anyone's time, but if you happen to have an idea as to what I am doing wrong I would greatly appreciate it!

**Comment by mikehardy** [comment]
Are you able to do builds to your real device connected to your computer? I don't have anything concrete on why your FCM token would be invalid but it might be that the config still has something wrong on the apple side. You may find something more useful in the firebase-ios-sdk issue tracker or on stack overflow though I understand those are sometimes fruitless fishing expeditions

If the device tokens aren't working then obviously there will be a problem. I definitely get tokens that work in my app using current stable versions of everything so while it isn't prescriptive in your case I can say current code *can* work, I've seen it, which makes me think the config is still off somehow

**Comment by michorvath** [comment]
Unfortunately I only have windows machines so I am using MacInCloud and Testflight for every build (I've tried existing usb over rdp software and none of it works for development). I will keep searching through stack and the issue tracker. I'm at least now aware that the issue is the tokens I am getting are apparently not registered, so something with Apple. I made an Android build and it worked on device. I'm close I hope 🤞

**Comment by mikehardy** [comment]
I'm not sure how they are doing it, but I can say conclusively that you really must use real devices for testing - if their setup is using simulators it will definitely not work, and when I scan their offerings I do not see any mention of real devices.

That might be your cause right there, as even if you get a token from a simulator, it will not work.

**Comment by michorvath** [comment]
Yes it is a cloud based mac solution for dev work. I am just making archives to test things and putting them on the app store in Testflight so I can test on actual devices. I'm pretty sure this would be a lot easy to debug if I didn't have to do that.

**Comment by georgetk1996** [comment]
Just a heads-up : So I was using `setBackgroundMeasageHandler` for saving a token from the payload to the device storage on iOS. It wasn't working at all on iOS, but on Android it worked seamlessly. After discussing with a native developer, he told me that Apple won't give control to the app, when it isn't in foreground.

So I have used another package - `PushNotificationIOS`, specifically the method [getDeliveredNotifications](https://github.com/react-native-community/push-notification-ios#getdeliverednotifications), to get all the notifications in the iOS device Notification Center, when the app comes back to foreground.

Basically, this method gives an array of notifications. In my case, I only needed the notification at array[0].

**Comment by mikehardy** [comment]
> After discussing with a native developer, he told me that Apple won't give control to the app, when it isn't in foreground.

This is imprecise. If you send a cloud message with data-only payload even, and content-available: 1 in the apple part of the payload, and you have the correct configuration for your app (background mode messaging), and the user uses your app enough while your app is reasonable enough on resources to be allocated some power budget, iOS will wake your app and your app will execute code. I do this very thing, it can work - not 100% reliable but it can work. And when your app comes foreground the messages will be delivered anyway, so it will work the same.

Additionally with react-native-background-fetch you can periodically get control in the background if the user interacts with your app semi-regularly in order for iOS to give it power budget. Similarly your tasks will run when it comes in foreground either way, so it will work the same. I do this very thing, it can work.

If your app uses the motion API or location APIs it may wake up as well, in the background.

There are several avenues to wake up apps in the background. They all have tradeoffs, but to say "Apple won't give control to the app, when it isn't in foreground", is definitely not correct.

**Comment by michorvath** [comment]
I have seen some conflicting instructions on setting up react-native-firebase for ios, so I just wanted to get a confirmation here: is it required that I add the APNs key file to my macs keychain for the app to be able to receive notifications? The file is already added to the Firebase project, as I am using the pre-existing project. I don't actually have the file and can't generate a new key due to limitations from the developer console so I contacted the previous developer to get the key, but is it even necessary locally for signing and all?

**Comment by mikehardy** [comment]
I wish I knew but that is a very precise question and all I can say is that when I set it up, I followed the instructions exactly from rnfirebase.io - originally the older v5 instructions (they are linked in the sidebar). Not sure what the mac keychain result of doing that is but I build on a quite a few macs and don't recall doing anything with the local keychain, also people use cloud CI poviders which I assume start with clean keychains

I do use fastlane to automate all of the iOS junk which installs the developer profiles into the keychain though, if I understand it's process correctly

**Comment by michorvath** [comment]
I ignored revoking and making a new APNs key to add to my keychain and instead completely removed Firebase from my project and readded following the docs exactly. I still have the same issue, Android works flawlessly and iOS gives me the same error when attempting to send a notification via Postman: "The registration token is not a valid FCM registration token". I know that v5 works with all the profiles and setup from the Apple developer console as the old app still technically functions. The Firebase project is using the APN key instead of APN certificates and my code contains the correct services files for both Android and iOS.

I have run out of things to try, I honestly have no clue. On past projects when I have used OneSignal or AWS to send push notifications I've never encountered an issue like this. It has taken 80+ hours of my time and I am not even sure at this point what could be causing the issue. If I google that error that I receive, there are only about 4 relevant articles related to Firebase v6 and none of them contain a viable solution. There really isn't much to setting up Firebase so I have no idea how others are getting it working but this project doesn't seem to want to work for iOS.

**Comment by mikehardy** [comment]
If it's definitely on a real device (because simulators do generate invalid tokens - remote messaging doesn't work on a simulator) then perhaps this related comment and workaround for an underlying firebase-ios-sdk bug is required: https://github.com/invertase/react-native-firebase/issues/4299#issuecomment-705128266

**Comment by michorvath** [comment]
I gave that a read and if I am understanding your comment correctly, I just add the "podfile chunk" from the bottom of that comment to my podfile then run `cd ios && rm -f Podfile.lock && pod install --repo-update`?

I'm sorry if I am making this more complicated... I am primarily a web developer and UX designer. A lot of this stuff is new to me.

**Comment by mikehardy** [comment]
That should be correct - those lines in your Podfile - then "reinstall pods" (and the commands I gave are mostly geared towards making sure the cocoapods definitions are up to date

**Comment by michorvath** [comment]
I added those lines to my podfile and reinstalled the pods with the commands you gave and also added code to check if the APNS key is returned on the device with `messaging().getAPNSToken()` and I still get the same error. I do get the APNS token as well as a FCM token, but when I attempt to send the test notification with the FCM token, I get the 400 error.

**Comment by michorvath** [comment]
So I plugged the APNS token I got from Firebase into [https://github.com/node-apn/node-apn](url) and attempted to send a production notification (as my app is in Testflight on a real device) and I get a "bad device token" error. I know the environment is not mismatched, both are set to production. Is there some misconfiguration that would cause my app to have both an invalid FCM and APNS token from using react-native-firebase?

**Comment by mikehardy** [comment]
I think there is an assumption there - that both the FCM and APNS token are valid. It could be that the APNS token is invalid, and the FCM is correct at the FCM level but transitively invalid via the APNS invalidity, if that makes sense. Stated differently, if the APNS token were valid then the FCM token would be fine and work. Why are you generating invalid APNS tokens? I'm not sure
